### DISCUSSION TITLE
Trying add DNS servers with new 5.9.9 on Fedora

### DISCUSSION THREAD

**Discussion by robdyck** [strongswan/strongswan#1528] [General]
I am using the resolve plugin to set the DNS servers for a road warrior. Fedora default installation use systemd-resolved to implement DNS. The command line tool to manage it is resolvectl. The resolve plugin uses the resolvconf tool to set DNS servers. On Fedora resolvconf is not a stand alone tool but is actually another way of accessing resolvectl.

The default interface, lo, in resolve.conf seems not to be a good default.
``
resolvconf -a lo < dns-add
Failed to set DNS configuration: Could not activate remote peer: activation request failed: unknown unit.
``

The next observation is that the original DNS servers are not restored when the SA is terminated.
We have no DNS servers until NetworkManager is restarted or use the network widget to disconnect and reconnect.

Also I was wondering if it was possible enhance the process so that when an SA is initiated and a DNS change is required, the process chooses the active interface. A road warrior might sometimes connect by WiFi and sometimes connect by tethering to a phone. Tethering results in an ephemeral interface being created. At the present time this would require editing resolve.conf.

**Comment by tobiasbrunner** [comment]
> The default interface, lo, in resolve.conf seems not to be a good default.

Yeah, that might not be ideal. The main issue is, of course, that `systemd-resolved` associates DNS servers with actual interfaces (instead of some arbitrary identifier).

> resolvconf -a lo < dns-add Failed to set DNS configuration: Could not activate remote peer: activation request failed: unknown unit.

That sounds like a weird error, though, if the problem is the interface ("activation" and "unknown unit" indicate fundamental issues with `systemd` and/or D-Bus). Does it work if you configure the actual physical interface?

> The next observation is that the original DNS servers are not restored when the SA is terminated.

You might have configured it incorrectly. Did you just configure the interface name in `resolvconf.iface`? It might work if you add a "protocol" value after a dot (e.g. `<iface>.ipsec`, like the default value). But I guess it depends on what the `resolvconf` implementation provided by `resolvectl` actually does.

> Also I was wondering if it was possible enhance the process so that when an SA is initiated and a DNS change is required, the process chooses the active interface.

I guess that would theoretically be possible. But we might not chose the correct interface, and if the client is mobile and the interface is later switched they won't be moved either (not sure if that makes a difference for `systemd-resolved` - again associating DNS servers with actual interfaces is just stupid).

I'd suggest you look into using an [XFRM interface](https://docs.strongswan.org/docs/5.9/features/routeBasedVpn.html#_xfrm_interfaces_on_linux) as that will allow you to statically configure that in `resolvconf.iface`. It's an approach that I've implemented for the NetworkManager backend as NM does something similar and with `dnsmasq` in an even worse way as it binds the latter to the interface (see https://github.com/strongswan/strongswan/issues/1048#issuecomment-1375825424 , no feedback yet on the patches by any users).

**Reply by robdyck** [reply]
I had a look at the resolvconf/resolvectl code. It initially connects with resolved's d-bus but results in some failure. There are no debug logs to say why it was failing. Anyway when that didn't work it tries to connect to the networkd d-bus. In Fedora systemd-networkd is not enabled. I think that is why I get the "unknown unit" error.

With an SA established there is no xfrm device
``resolvconf:Failed to resolve interface "xfrm": No such device
``

What misconfiguration would cause the original name servers not be restored?

**Reply by tobiasbrunner** [reply]
> I had a look at the resolvconf/resolvectl code. It initially connects with resolved's d-bus but results in some failure. There are no debug logs to say why it was failing. Anyway when that didn't work it tries to connect to the networkd d-bus. In Fedora systemd-networkd is not enabled. I think that is why I get the "unknown unit" error.

I see.

> With an SA established there is no xfrm device

There won't be automatically, you have to create one and then reference it in the config via `if_id_*` (see [the docs](https://docs.strongswan.org/docs/5.9/features/routeBasedVpn.html#_xfrm_interfaces_on_linux)). You also have to install routes via that interface manually (unless you use the patches from the NM branch), either before or via updown/vici script. And if the IKE traffic overlaps with those routes you have to either install a direct route to the VPN server via physical interface, or configure marks for the IKE socket and strongSwan's routing table like the NM patches do (you'd have to install the routes in that table to make it work).

> What misconfiguration would cause the original name servers not be restored?

We call `resolvconf` with `-a <iface.protocol>` to install the servers and then call it again with `-d <iface.protocol>` to remove them. With classic `resolvconf` that removes only those added with that identifier later (so e.g. servers added with `<iface>.dhcp` would not be affected). That's why that protocol part should be added to "namespace" the servers. However, if the `resolvectl` implementation just associates the added DNS servers with the interface (i.e. if it basically ignores the `.protocol` part) and then simply removes all of them from that interface config later, it obviously won't be able to restore the previous servers. Via `resolvconf` we also don't have the ability to query any existing servers to restore them later ourselves. I guess that would require a D-Bus client that directly accesses `org.freedesktop.resolve1`, or, as suggested above, a dedicated interface that's only used for IPsec. Or could `openresolv` be an option?

**Reply by robdyck** [reply]
I did some reading about XFRM interface and it wouldn't be useful in my use case. Yes, it is not automatic. Also it has to be associated with a real interface so I can't use it like some generic interface in resolve.conf. Furthermore strongswan would no longer setup the routing for me.

Off on a tangent here, the desired DNS servers are not setting the environment variables that are passed to _updown.

**Reply by tobiasbrunner** [reply]
> Also it has to be associated with a real interface

It does not. The interface can be omitted (at least with newer versions of iproute2 and the kernel), and even if you set one, it doesn't really matter in practice.

> Furthermore strongswan would no longer setup the routing for me.

No, you have full flexibility how/what to route via VPN (and you could use the mentioned patches to let strongSwan automatically install routes).

> Off on a tangent here, the desired DNS servers are not setting the environment variables that are passed to _updown.

They are not, but how exactly would that help?

**Reply by robdyck** [reply]
Could you clarify this statement from the strongswan documentation.
``
The daemon will not install any routes for CHILD_SAs with outbound interface ID, so it’s not necessary to disable the route installation globally.
``
With regard to the DNS environment variables if they were passed to _updown then DNS could be managed by the script.

**Reply by tobiasbrunner** [reply]
> Could you clarify this statement from the strongswan documentation.

It means it's not necessary to disable route installation globally via `charon.install_routes` in strongswan.conf (as compared to e.g. with VTI interfaces).

> With regard to the DNS environment variables if they were passed to _updown then DNS could be managed by the script.

Yeah, but what would you do differently with them? Or are you considering creating an XFRM interface dynamically?

**Reply by robdyck** [reply]
Yes I thought about creating the XFRM interface in the script.
It seems odd that variable names are created for DNS in the environment but they have no value set.
``
PLUTO_DNS4_$i
contains an IPv4 DNS server attribute received from a responder, $i enumerates from 1 to the number of servers per address family

PLUTO_DNS6_$i
contains an IPv6 DNS server attribute received from a responder, $i enumerates from 1 to the number of servers per address family
``
Openresolv is working for me but I may investigate xfrm interace further and look at the patch mentioned.

**Reply by tobiasbrunner** [reply]
> It seems odd that variable names are created for DNS in the environment but they have no value set.

They are only set with specific configurations (also see #1320). That might change in the future, though.

**Comment by robdyck** [comment]
Associating DNS with an interface might be useful if you more than one active interface.

**Reply by tobiasbrunner** [reply]
Not really. You need to know the target IP address (i.e. perform a DNS query) before you can do a route lookup that decides on which interface the packet goes out.

Associating them with the source (DHCP, VPN client, manual config) would make way more sense (perhaps with the option for users to prefer some sources over others), plus some arbitrary identifier so that multiple servers etc. can be removed later in bulk.

**Comment by robdyck** [comment]
I tried adding a suffix to the interface name in resolve.conf but it makes no difference.

**Comment by tibz7** [comment]
I'm having the same problem, did someone figureout a fix?

**Reply by tibz7** [reply]
i just found a fix: I added `iface=ipsec0` in `/etc/strongswan/strongswan.d/charon/resolve.conf`

**Reply by robdyck** [reply]
As I wrote some months ago openresolv worked for me. Fedora's version of resolvconf is not compatible with systemd-resolved when used with strongswan. At the time I said I would look into using an xfrm interface but I never did. I assume that is how you created your ipsec0. When I upgraded Fedora the link at /usr/bin/resolvconf to /etc/alternatives was removed thus dsiabling openresolv. I saw your note and it prompted me to try creating and xfrm interface as per strongswan's example and changing swanctl.conf's child configuration to use it. So it worked for me. The road warrior was able to set the DNS server as per the instruction it received with IKE.

I hadn't actually needed the road warrior connection since the Fedora upgrade but your note prompted me to look at an alternative that wouldn't be disabled every time Fedora was upgraded.
